content/helpers.go

```diff
@@ -80,6 +81,21 @@ func WriteBlob(ctx context.Context, cs Ingester, ref string, r io.Reader, desc o
 	return Copy(ctx, cw, r, desc.Size, desc.Digest, opts...)
 }
 
+// WriteLayer writes a layer's data into the content store.
+//
+// This is useful when the size and digest are NOT known beforehand.
+func WriteLayer(ctx context.Context, cs Store, r ReaderDigester, opts ...Opt) (int64, digest.Digest, error) {
```

Comment:
Let's try to avoid the new interface. Also, notice the rest of this package doesn't use the word "layer". The other option is to handle the case where `desc` has an empty digest and size. Also, this should use `Ingester` over `Store`, since we aren't reading blobs back from the argument.

Reply:
Also, looks like Copy handles missing size and digest: https://github.com/containerd/containerd/blob/master/content/helpers.go#L123.

Reply:
What gives me the size and digest *after* the layer 'transformation' (encryption or decryption) so that I can fill out the descriptor of the layer so that I can write it into the manifest? The transformation happens while reading for example a plain layer from the content store (whose hash (=filename) and size we obviously have).  The digest and size of the encrypted layer is then not the same as the digest and size of the plain layer that was originally read from the content store... I don't think there is anything that hashes (using a resumable hash) the data while copying them since you normally have them anyway.

Reply:
What was the issue is the fact that io.CopyBuffer (https://github.com/containerd/containerd/blob/master/content/helpers.go#L205) takes a reader as an input for the source and we have no way of hashing along the transformed data (we read the plain data and encrypt them while the Reader is Read() from) and determining the number of bytes outside that API  unless we were now to change this and read chunks from the Reader ourselves and hash along and pass in the byte array for copying purposes.

Reply:
To summarize the problem: We have to be able to hash along the data we are transforming (and also determine the size). Two choices here where to do this: 
- Inside the Reader implementation the data transformation; this is the path I chose with the interface
- Re-implement the function io.CopyBuffer() and hash along the chunks we are Read()ing from the Reader.

Reply:
Would getting the size and digest from the writer work? Kind of like:
```
    err := Copy(ctx, cw, r, 0, "", opts...)
    if err != nil {
      return 0, "", err
    }
    return cw.Status().Total , cw.Digest(), nil
```

Reply:
So, yes, the Writer is already hashing it, so we can use it.

Can we change the return parameters from existing WriteBlob() to include size and Digest or do we have to create a new function?

Reply:
pushed a patch that makes use of the Writer-side hash.